README.md
Welcome to Zandronum!
=====================

A continuation of Skulltag.
---------------------------

Zandronum is a multiplayer oriented port, based off Skulltag, for Doom and Doom II by [id Software](http://www.idsoftware.com/).

Zandronum brings classic Doom into the 21st century, maintaining the essence of what has made Doom great for so many years and, at the same time, adding new features to modernize it, creating a fresh, fun new experience.

Here's why you should kill your time with Zandronum, versus vanilla Doom:
 
* Excellent online play
    * Client/server architecture: play and switch between games with ease
    * Supports up to 64 players
        * Wide array of gameplay modes
        * Competitive: Team DM, Duel, CTF, Possession, LMS, Terminator, Skulltag
        * Cooperative: Survival, Invasion (or both!)
        * Modifiers, like Instagib and Buckshot, can be used to spice up any game.
* Modern and beautiful
    * Support for large number of [ZDoom](http://zdoom.org) and [GZDoom](http://www.osnanet.de/c.oelckers/gzdoom/index.html) mods
    * Choose between GZDoom's OpenGL renderer, or Doom's Software renderer
    * Useful features like a console, Quake-style key bindings, freelooking, and jumping
    * Support for many Doom engine based games including Heretic, Hexen, and Strife
    * Supports modern versions of Windows, Mac OS X, and Linux.
* Compatible with Skulltag mods
    * Supports announcers and skins for online play
    * Runes, new weapons, and new monsters for modders
    * Bots, and support for custom bots
* So much more!

Links
-----

* [Official Website](http://zandronum.com/)
* [Community Forums](http://zandronum.com/forum/)
* [Wiki](http://wiki.zandronum.com/)
* [Bugtracker](http://zandronum.com/tracker/)


License
-------

New code in Zandronum is released under a 4-clause license based on the OSI-approved and GPL-compatible [Sleepycat License](http://www.opensource.org/licenses/Sleepycat), with the addition of the "No Endorsement" clause from the 3-clause [New BSD License](http://www.opensource.org/licenses/BSD-3-Clause).  Practically, this means that Zandronum's source code is safe to use in either GPL or Doom Source License/Raven Source License/etc. source ports as long as the terms of the Zandronum license are satisfied.

The Zandronum license only covers Zandronum-specific code.  Some source files contain additional notices of original copyright by their contributors.

See [ZDoom license page](http://zdoom.org/wiki/license) for more details on other licenses used within Zandronum.


dumb/ChangeLog
======================= foo_dumb part =================================

2007-01-26 23:42 UTC - kode54
- Added CPUID specialization for it_filter() so I can release it.

2007-01-26 22:50 UTC - kode54
- Moved IT S70-2 effects alongside the rest of S7x so they all trigger after
  paired notes.
- Integrated note on/cut ramping with volume transition ramping to reduce
  setup/cleanup overhead of calling render_playing and the resampler functions
  for single samples.
- Note on/cut ramping scaled from 7 sample logarithmic and 256 sample linear to
  .75ms and 5ms respectively, both linear.
- Integrated the pattern looping changes from DUMB 0.9.3, since a few files
  seemed to be broken. The XM-only features still needed to be flagged for XM
  so they won't break MOD.
- Implemented SSE replacement for it_filter() which is actually twice as fast.

2007-01-24 09:37 UTC - kode54
- Version is now 0.9.8.3

2007-01-24 08:34 UTC - kode54
- Implemented 669 running effects correctly.

2007-01-23 17:42 UTC - kode54
- Fixed IT reader decompression to support compression+delta on files created
  with 2.15 or newer.

2007-01-22 18:54 UTC - kode54
- Implemented 669 running effects, nasty stuff.

2007-01-22 16:40 UTC - kode54
- Fixed MOD loading step in case any files fall through another loader first.
  (if asked twice.mod)

2007-01-21 19:47 UTC - kode54
- Implemented crappy fixed end of sample volume ramping.

2007-01-21 17:43 UTC - kode54
- XM reader now ignores sample header length, as Sk@le Tracker fills this with
  nonsense.

2007-01-21 14:21 UTC - kode54
- Fixed a bug in IT renderer / get_true_pan that caused it to crash on IT files
  with instruments enabled upon hitting an invalid instrument change.

2006-12-26 06:41 UTC - kode54
- Fixed sample rate reporting for when extra dynamic info is turned off.

2006-11-01 00:50 UTC - kode54
- Resamplers now return immediately if the fixed point delta is zero.
- IT renderer now cuts existing notes when it reaches a note where the instrument
  references an invalid sample. (jumping under trees.it)
- Version is now 0.9.8.2

2006-10-14 08:33 UTC - kode54
- IT renderer pattern loop processor now resets the rows played to clear table
  per channel regardless of whether order changed since last loop start was
  initiated. This is correct, since two pattern loop starts in a row means that
  a loop will never jump to the first loop point.

2006-10-12 06:28 UTC - kode54
- STM reader checks for more values after the song name, reports the correct
  order list size, and ignores samples which are shorter than four bytes or
  have a default volume of zero.

2006-10-08 03:51 UTC - kode54
- Added CP437 graphics to text conversion functions.

2006-10-07 05:24 UTC - kode54
- Simplified volume ramping update code, reducing the size of the resampler code
  considerably.
- Bumped the volume ramping precision to 24 bits of fraction precision, which is
  needed by Sweetsin.xm.
- Moved sample rate reporting to dynamic info as requested by Peter, since it's
  not a property of the files themselves, but user configurable.
- Version is now 0.9.8.1

2006-10-07 03:42 UTC - kode54
- Changed DSMF sample loader to ignore unknown flags instead of blowing an error.

2006-09-25 17:39 UTC - kode54
- Added hack to MOD loader for when sample start is specified in bytes instead of
  words.

2006-09-19 15:05 UTC - kode54
- Shuffled finetune calculation into the correct position, immediately applied
  as delta is calculated from note.
- Promoted IT_SAMPLE finetune property to signed short as char is insufficient
  for full semitone range. (+/- 256)
- Changed resampler to use full 64-bit comparison for todo variable range checking
  which should hopefully eliminate any further problems with pitch slides which go
  out of range.
- Version is now 0.9.8

2006-06-16 19:10 UTC - kode54
- Fixed MOD loader FLT8 handler so it halves the order list after it's allocated
  and loaded.
- Unified renderer and loop restarting for playback and seeking, and made seeking
  abortable.

2006-06-16 09:12 UTC - kode54
- Fixed PSM subsong info reporting.

2006-06-15 21:23 UTC - kode54
- Added finetune property to IT_SAMPLE, modified ASY, MOD, MTM, old PSM, and XM
  to use it instead of adjusting the C5 speed, so that finetune effect overrides
  rather than compounding sample finetune.

2006-06-10 18:10 UTC - kode54
- Fixed volume ramping.

2006-06-09 16:48 UTC - kode54
- Ported most of DUMB 0.9.3 over, except for the conflicting pattern loop change

2006-06-03 02:13 UTC - kode54
- Implemented ASYLUM Music Format support.

2006-06-02 22:06 UTC - kode54
- Fixed finite looping for speed 0 and global volume 0 triggers.
- Reset loop count when seeking backwards.

2006-06-02 21:20 UTC - kode54
- IT renderer now supports pattern jump and break to row on the same row.
  (coffee-blues 3.mod)
- Version is now 0.9.7.5

2006-06-02 20:25 UTC - kode54
- Empty pattern scanner now returns the correct value for completely empty
  pattern structures, and also checks for lists of empty rows.

2006-06-01 00:47 UTC - kode54
- STM reader reads effects properly now.

2006-05-31 23:15 UTC - kode54
- Implemented STM support.
- Version is now 0.9.7.4

2006-05-31 20:12 UTC - kode54
- Finished implementing working fading support.

2006-05-31 19:09 UTC - kode54
- Implemented fading support.

2006-05-31 18:31 UTC - kode54
- Removed context dialog from resources.

2006-05-30 03:57 UTC - kode54
- IT reader works around a broken file that has several of the instruments
  offset by two bytes from their indicated file offsets. Since the hack
  only checks for offset by two null bytes, it also assumes that the two
  bytes it can't read from the end are also zero. (bz_ult9.it)
- IT reader treats null instrument/sample/pattern offsets as empty items.
  (flight2.it, sherri.it)

2006-05-30 00:04 UTC - kode54
- XM reader now truncates instrument envelope lengths to 12, and performs
  some basic envelope loop checking, similar to Open ModPlug Tracker.

src/d_gui.h
/*
** d_gui.h
**
**---------------------------------------------------------------------------
** Copyright 1998-2006 Randy Heit
** All rights reserved.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions
** are met:
**
** 1. Redistributions of source code must retain the above copyright
**    notice, this list of conditions and the following disclaimer.
** 2. Redistributions in binary form must reproduce the above copyright
**    notice, this list of conditions and the following disclaimer in the
**    documentation and/or other materials provided with the distribution.
** 3. The name of the author may not be used to endorse or promote products
**    derived from this software without specific prior written permission.
**
** THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
** IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
** OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
** IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
** INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
** NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
** DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
** THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
** (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
** THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
**---------------------------------------------------------------------------
**
** So when do I get a real UT-like windowing system?
*/

#ifndef __D_GUI_H__
#define __D_GUI_H__

// For a GUIEvent, x and y specify absolute location of mouse pointer
enum EGUIEvent
{
	EV_GUI_None,
	EV_GUI_KeyDown,			// data1: unshifted ASCII, data2: shifted ASCII, data3: modifiers
	EV_GUI_KeyRepeat,		// same
	EV_GUI_KeyUp,			// same
	EV_GUI_Char,			// data1: translated character (for user text input), data2: alt down?
	EV_GUI_FirstMouseEvent,
		EV_GUI_MouseMove,
		EV_GUI_LButtonDown,
		EV_GUI_LButtonUp,
		EV_GUI_LButtonDblClick,
		EV_GUI_MButtonDown,
		EV_GUI_MButtonUp,
		EV_GUI_MButtonDblClick,
		EV_GUI_RButtonDown,
		EV_GUI_RButtonUp,
		EV_GUI_RButtonDblClick,
		EV_GUI_WheelUp,			// data3: shift/ctrl/alt
		EV_GUI_WheelDown,		// "
		EV_GUI_WheelRight,		// "
		EV_GUI_WheelLeft,		// "
		EV_GUI_BackButtonDown,
		EV_GUI_BackButtonUp,
		EV_GUI_FwdButtonDown,
		EV_GUI_FwdButtonUp,
	EV_GUI_LastMouseEvent,
};

enum GUIKeyModifiers
{
	GKM_SHIFT	= 1,
	GKM_CTRL	= 2,
	GKM_ALT		= 4,
	GKM_LBUTTON = 8
};

// Special codes for some GUI keys, including a few real ASCII codes.
enum ESpecialGUIKeys
{
	GK_PGDN		= 1,
	GK_PGUP		= 2,
	GK_HOME		= 3,
	GK_END		= 4,
	GK_LEFT		= 5,
	GK_RIGHT	= 6,
	GK_ALERT	= 7,		// ASCII bell
	GK_BACKSPACE= 8,		// ASCII
	GK_TAB		= 9,		// ASCII
	GK_LINEFEED	= 10,		// ASCII
	GK_DOWN		= 10,
	GK_VTAB		= 11,		// ASCII
	GK_UP		= 11,
	GK_FORMFEED	= 12,		// ASCII
	GK_RETURN	= 13,		// ASCII
	GK_F1		= 14,
	GK_F2		= 15,
	GK_F3		= 16,
	GK_F4		= 17,
	GK_F5		= 18,
	GK_F6		= 19,
	GK_F7		= 20,
	GK_F8		= 21,
	GK_F9		= 22,
	GK_F10		= 23,
	GK_F11		= 24,
	GK_F12		= 25,
	GK_DEL		= 26,
	GK_ESCAPE	= 27,		// ASCII
	GK_FREE1	= 28,
	GK_FREE2	= 29,
	GK_BACK		= 30,		// browser back key
	GK_CESCAPE	= 31		// color escape
};

#endif //__D_GUI_H__
